### ISSUE TITLE
Reference whole MapMessage - and not just keys

### ISSUE THREAD

**Issue by magnuskvalheim** [vy/log4j2-logstash-layout#68]
I would like to add the content of MapMessage as a separate field.
Currently it's only possible to reference named fields, but ours are dynamic so would like to add all key-value pairs in a field for that.
Example: 
`<KeyValuePair key="customfield" value="${json:map}"/>`

A workaround could be to also add the mapmessage fields to mdc, but like to avoid add it both places + removing from mdc once the entry is logged.

So instead of
```
try (final CloseableThreadContext.Instance ctc = CloseableThreadContext.putAll(params)) {
	log.info(new StringMapMessage(params));
}
```
I'd rather do
`log.info(new StringMapMessage(params));`

Another possibility is to use the message:json, but that has the disadvantage of also outputting non MapMessages `({"message": <formattedMessage>} object)`

Is it possible to reference the map without key, like it's possible with mdc?

Br

**Comment by magnuskvalheim** [comment]
Hi again.

The ecs-logging-java project supports structured logging by adding the MapMessage to the document root.
I'm thinking that could be a useful feature. Is that something you may have considered as well? 

1)Perhaps by adding a parameter to LogstashLayout. 
2)..or maybe better - as a <Unwrapped value="" /> to the <EventTemplateAdditionalFields> ?

If You think it's a reasonable scenario I could try to take a stab at implementing it.

**Comment by vy** [comment]
Hey @magnuskvalheim! Sorry for the late response, I did not miss about it. LogstashLayout successor JsonTemplateLayout will be released this week or so. It adresses your issue by supporting merging of an object with the parent. This week I will see if I can backport it to LogstashLayout.

**Comment by magnuskvalheim** [comment]
Hi @vy . 
Thanks for reply. 
Ok, that sounds great - I'll test that out once it becomes available 👍

**Comment by vy** [comment]
@magnuskvalheim, Log4j 2.14.0 is released. Would you mind giving it a try, please?

**Comment by magnuskvalheim** [comment]
Hi @vy 
Great - thanks for your hard work.
I'll try it out and report back :)

**Comment by magnuskvalheim** [comment]
Hi @vy 

Thanks and congratulations with the release!

I've tried the Log4j 2.14.0 release and the flattening feature of mdc resolver - that works really well!
Further - I wanted to have the contents of MapMessage in a custom field - and that is possible with reusing the message resolver.
So amazing job there!

With the message resolver it's possible to set the fallbackKey - an item to wishlist would be to be able to do the opposite. That is option to disregard SimpleMessage (as it's already printed in the 'message' field.) Another option could be to use the mapresolver without specifying key?

--
I also tried the EventTemplateAdditionalFields and saw some issues on my setup.
Example
```
<JsonTemplateLayout>
	<EventTemplateAdditionalFields>
		<EventTemplateAdditionalField key="host" value="www.apache.org"/>
		<EventTemplateAdditionalField key="_serviceName" value="auth-service"/>
		<EventTemplateAdditionalField key="_containerId" value="6ede3f0ca7d9"/>
	</EventTemplateAdditionalFields>
</JsonTemplateLayout>
```
Is giving me exceptions:
```
2020-11-12 22:33:47,236 main ERROR Could not create plugin of type class org.apache.logging.log4j.layout.template.json.JsonTemplateLayout$EventTemplateAdditionalField for element EventTemplateAdditionalField: java.lang.IllegalArgumentException: blank key java.lang.IllegalArgumentException: blank key
	at org.apache.logging.log4j.layout.template.json.JsonTemplateLayout$EventTemplateAdditionalField$Builder.validate(JsonTemplateLayout.java:676)
	at org.apache.logging.log4j.layout.template.json.JsonTemplateLayout$EventTemplateAdditionalField$Builder.build(JsonTemplateLayout.java:670)
	at org.apache.logging.log4j.layout.template.json.JsonTemplateLayout$EventTemplateAdditionalField$Builder.build(JsonTemplateLayout.java:644)
	at org.apache.logging.log4j.core.config.plugins.util.PluginBuilder.build(PluginBuilder.java:122)
	at org.apache.logging.log4j.core.config.AbstractConfiguration.createPluginObject(AbstractConfiguration.java:1002)
	...

2020-11-12 22:33:47,237 main ERROR Unable to invoke factory method in class org.apache.logging.log4j.layout.template.json.JsonTemplateLayout$EventTemplateAdditionalField for element EventTemplateAdditionalField: java.lang.IllegalStateException: No factory method found for class org.apache.logging.log4j.layout.template.json.JsonTemplateLayout$EventTemplateAdditionalField java.lang.IllegalStateException: No factory method found for class org.apache.logging.log4j.layout.template.json.JsonTemplateLayout$EventTemplateAdditionalField
	at org.apache.logging.log4j.core.config.plugins.util.PluginBuilder.findFactoryMethod(PluginBuilder.java:234)
	at org.apache.logging.log4j.core.config.plugins.util.PluginBuilder.build(PluginBuilder.java:134)
	...

2020-11-12 22:33:47,238 main ERROR EventTemplateAdditionalField contains invalid attributes "value", "key"
2020-11-12 22:33:47,238 main ERROR Could not create plugin of type class org.apache.logging.log4j.layout.template.json.JsonTemplateLayout$EventTemplateAdditionalField for element EventTemplateAdditionalField: java.lang.IllegalArgumentException: blank key java.lang.IllegalArgumentException: blank key
	at org.apache.logging.log4j.layout.template.json.JsonTemplateLayout$EventTemplateAdditionalField$Builder.validate(JsonTemplateLayout.java:676)
	at org.apache.logging.log4j.layout.template.json.JsonTemplateLayout$EventTemplateAdditionalField$Builder.build(JsonTemplateLayout.java:670)
	at org.apache.logging.log4j.layout.template.json.JsonTemplateLayout$EventTemplateAdditionalField$Builder.build(JsonTemplateLayout.java:644)
	at org.apache.logging.log4j.core.config.plugins.util.PluginBuilder.build(PluginBuilder.java:122)
	...
2020-11-12 22:33:48,911 main ERROR Null object returned for EventTemplateAdditionalField in EventTemplateAdditionalFields.
2020-11-12 22:33:48,911 main ERROR Null object returned for EventTemplateAdditionalField in EventTemplateAdditionalFields.
2020-11-12 22:33:48,912 main ERROR Null object returned for EventTemplateAdditionalField in EventTemplateAdditionalFields.
```

**Comment by vy** [comment]
@magnuskvalheim, thanks so much for taking time to try it out and your kind words! Would you mind creating two tickets (assigned to my name) in [the Log4j2 JIRA board](https://issues.apache.org/jira/projects/LOG4J2) for these, please?

**Comment by vy** [comment]
Created https://issues.apache.org/jira/browse/LOG4J2-2961

**Comment by magnuskvalheim** [comment]
Great - thanks for addressing it so quickly.
If you want to I guess you can close this ticket now. I'm following the log4j2 issues.

**Comment by unoexperto** [comment]
@magnuskvalheim @vy Guys, could you please show example of json config with working `KeyValuePair` ? It drives me crazy :(

https://issues.apache.org/jira/browse/LOG4J2-3013

This is my complete config https://gist.github.com/unoexperto/faa5e251a478e567bbad2c5d92460053

**Comment by vy** [comment]
@unoexperto, I have shared [one in LOG4J2-3013](https://issues.apache.org/jira/browse/LOG4J2-3013?focusedCommentId=17277895&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17277895).

**Comment by pippy44** [comment]
@vy Hi I see that both https://issues.apache.org/jira/browse/LOG4J2-2961 and https://issues.apache.org/jira/browse/LOG4J2-3013 have been closed and marked as duplicates of https://issues.apache.org/jira/browse/LOG4J2-2995. Is the eventTemplateAdditionalFields issue being addressed in 2995? If so, is a 2.14.1 release expected soon?

**Comment by vy** [comment]
@pippy44, yes. I would recommend you raising your hand in Log4j `dev` mailing list as a user waiting for the 2.14.1 release.
